File: app/src/main/java/li/klass/fhem/service/room/RoomListService.java
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.service.room;

import android.content.Context;
import android.content.Intent;

import com.google.common.base.Optional;
import com.google.common.collect.Sets;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;

import javax.inject.Inject;
import javax.inject.Singleton;

import li.klass.fhem.appwidget.service.AppWidgetUpdateService;
import li.klass.fhem.constants.Actions;
import li.klass.fhem.constants.BundleExtraKeys;
import li.klass.fhem.constants.PreferenceKeys;
import li.klass.fhem.dagger.ForApplication;
import li.klass.fhem.domain.FHEMWEBDevice;
import li.klass.fhem.domain.core.Device;
import li.klass.fhem.domain.core.DeviceType;
import li.klass.fhem.domain.core.RoomDeviceList;
import li.klass.fhem.service.AbstractService;
import li.klass.fhem.service.SharedPreferencesService;
import li.klass.fhem.service.connection.ConnectionService;
import li.klass.fhem.service.intent.DeviceIntentService;
import li.klass.fhem.service.intent.NotificationIntentService;
import li.klass.fhem.service.intent.RoomListUpdateIntentService;
import li.klass.fhem.util.ApplicationProperties;

import static com.google.common.collect.Lists.newArrayList;
import static java.util.Collections.sort;
import static li.klass.fhem.constants.Actions.DISMISS_EXECUTING_DIALOG;
import static li.klass.fhem.constants.Actions.REDRAW_ALL_WIDGETS;
import static li.klass.fhem.constants.BundleExtraKeys.DO_REFRESH;
import static li.klass.fhem.constants.BundleExtraKeys.UPDATE_PERIOD;
import static li.klass.fhem.domain.core.DeviceType.AT;
import static li.klass.fhem.domain.core.DeviceType.getDeviceTypeFor;
import static li.klass.fhem.util.DateFormatUtil.toReadable;

@Singleton
public class RoomListService extends AbstractService {

    private static final Logger LOG = LoggerFactory.getLogger(RoomListService.class);

    public static final String PREFERENCES_NAME = RoomListService.class.getName();

    public static final String LAST_UPDATE_PROPERTY = "LAST_UPDATE";

    public static final long NEVER_UPDATE_PERIOD = 0;
    public static final long ALWAYS_UPDATE_PERIOD = -1;

    public static final String SORT_ROOMS_DELIMITER = " ";

    private final AtomicBoolean remoteUpdateInProgress = new AtomicBoolean(false);

    private List<Intent> resendIntents = newArrayList();

    volatile RoomDeviceList deviceList;

    @Inject
    ConnectionService connectionService;

    @Inject
    DeviceListParser deviceListParser;

    @Inject
    ApplicationProperties applicationProperties;

    @Inject
    @ForApplication
    Context applicationContext;

    @Inject
    SharedPreferencesService sharedPreferencesService;

    @Inject
    RoomListHolderService roomListHolderService;

    public void parseReceivedDeviceStateMap(String deviceName, Map<String, String> updateMap,
                                            boolean vibrateUponNotification) {

        Optional<Device> deviceOptional = getDeviceForName(deviceName);
        if (!deviceOptional.isPresent()) {
            return;
        }

        Device device = deviceOptional.get();
        deviceListParser.fillDeviceWith(device, updateMap);

        LOG.info("parseReceivedDeviceStateMap()  : updated {} with {} new values!", device.getName(), updateMap.size());

        applicationContext.startService(new Intent(Actions.NOTIFICATION_TRIGGER)
                .setClass(applicationContext, NotificationIntentService.class)
                .putExtra(BundleExtraKeys.DEVICE_NAME, deviceName)
                .putExtra(BundleExtraKeys.DEVICE, device)
                .putExtra(BundleExtraKeys.UPDATE_MAP, (Serializable) updateMap)
                .putExtra(BundleExtraKeys.VIBRATE, vibrateUponNotification));

        boolean updateWidgets = applicationProperties.getBooleanSharedPreference(PreferenceKeys.GCM_WIDGET_UPDATE, false);
        if (updateWidgets) {
            sendBroadcastWithAction(REDRAW_ALL_WIDGETS);

            applicationContext.startService(new Intent(REDRAW_ALL_WIDGETS)
                    .setClass(applicationContext, DeviceIntentService.class));
        }
    }

    /**
     * Looks for a device with a given name.
     *
     * @param deviceName name of the device
     * @return found device or null
     */
    @SuppressWarnings("unchecked")
    public <T extends Device<T>> Optional<T> getDeviceForName(String deviceName) {
        return Optional.fromNullable((T) getAllRoomsDeviceList().getDeviceFor(deviceName));
    }

    /**
     * Retrieves a {@link RoomDeviceList} containing all devices, not only the devices of a specific room.
     * The room device list will be a copy of the actual one. Thus, any modifications will have no effect!
     *
     * @return {@link RoomDeviceList} containing all devices
     */
    public RoomDeviceList getAllRoomsDeviceList() {
        RoomDeviceList originalRoomDeviceList = getRoomDeviceList();
        return new RoomDeviceList(originalRoomDeviceList);
    }

    /**
     * Loads the currently cached {@link li.klass.fhem.domain.core.RoomDeviceList}. If the cached
     * device list has not yet been loaded, it will be loaded from the cache object.
     * <p/>
     * <p>Watch out: Any modifications will be saved within the internal representation. Don't use
     * this method from client code!</p>
     *
     * @return Currently cached {@link li.klass.fhem.domain.core.RoomDeviceList}.
     */
    public RoomDeviceList getRoomDeviceList() {

        if (deviceList == null) {
            deviceList = roomListHolderService.getCachedRoomDeviceListMap();
        }
        return deviceList;
    }

    public void resetUpdateProgress() {
        LOG.debug("resetUpdateProgress()");
        remoteUpdateInProgress.set(false);
        resendIntents = newArrayList();
        sendBroadcastWithAction(DISMISS_EXECUTING_DIALOG);
    }

    /**
     * Method will check if a remote update of the current device list is required. This is
     * determined by two indicators:
     * <ul>
     * <li>After loading the cached device map, the cached map is still null. Effectively
     * this means that no devices had been cached.</li>
     * <li>The update period indicates that we have to update the device map.</li>
     * </ul>
     * <p/>
     * <p>
     * When finding out that we have to remotely update the device list, the current request
     * (as intent) is cached and an intent to {@link li.klass.fhem.service.intent.RoomListUpdateIntentService}
     * is sent. When the remote update completes, we will get an answer effectively calling
     * {#remoteUpdateFinished}. The method will resent all cached intents, resulting in
     * answers to waiting requests.
     * </p>
     * <p>
     * By caching calling intents that all want to remotely load device lists, we make
     * sure that only one remote update is concurrently executed. Also, we do not queue
     * remote requests, as they would load the same content from the server.
     * </p>
     *
     * @param intent       calling intent (that might request a remote update)
     * @param updatePeriod update period (that might indicate a necessary remote update)
     * @return {@link li.klass.fhem.service.room.RoomListService.RemoteUpdateRequired#REQUIRED} if the
     * calling intent will be cached and resend when the remote update has completed, otherwise
     * {@link li.klass.fhem.service.room.RoomListService.RemoteUpdateRequired#NOT_REQUIRED}
     */
    public RemoteUpdateRequired updateRoomDeviceListIfRequired(Intent intent, long updatePeriod) {
        if (deviceList == null) {
            deviceList = roomListHolderService.getCachedRoomDeviceListMap();
        }

        boolean requiresUpdate = shouldUpdate(updatePeriod) || deviceList == null;
        if (requiresUpdate) {
            LOG.info("updateRoomDeviceListIfRequired() - requiring update, add pending action: {}", intent.getAction());
            resendIntents.add(createResendIntent(intent));
            if (remoteUpdateInProgress.compareAndSet(false, true)) {
                applicationContext.startService(new Intent(Actions.DO_REMOTE_UPDATE)
                        .setClass(applicationContext, RoomListUpdateIntentService.class));
            }
            LOG.debug("updateRoomDeviceListIfRequired() - remote update is required");
            return RemoteUpdateRequired.REQUIRED;
        } else {
            LOG.debug("updateRoomDeviceListIfRequired() - remote update is not required");
            return RemoteUpdateRequired.NOT_REQUIRED;
        }
    }

    /**
     * Entry point for completed remote updates. See {@link #updateRoomDeviceListIfRequired} for
     * details on the process.
     */
    public void remoteUpdateFinished() {
        try {
            LOG.info("remoteUpdateFinished() - starting after actions");

            deviceList = roomListHolderService.getCachedRoomDeviceListMap();

            for (Intent resendIntent : resendIntents) {
                resend(resendIntent);
            }
            resendIntents.clear();

            LOG.info("remoteUpdateFinished() - requesting redraw of all appwidgets");

            applicationContext.sendBroadcast(new Intent(Actions.REDRAW_ALL_WIDGETS));
            applicationContext.startService(new Intent(Actions.REDRAW_ALL_WIDGETS)
                    .setClass(applicationContext, AppWidgetUpdateService.class)
                    .putExtra(BundleExtraKeys.ALLOW_REMOTE_UPDATES, false));

            LOG.info("remoteUpdateFinished() - remote update finished, device list is {}");
        } finally {
            LOG.info("remoteUpdateFinished() - finished, dismissing executing dialog");
            resetUpdateProgress();
        }
    }

    private void resend(Intent intent) {
        LOG.info("resend() : resending {}", intent.getAction());
        applicationContext.startService(createResendIntent(intent));
    }

    private Intent createResendIntent(Intent intent) {
        Intent resendIntent = new Intent(intent);
        resendIntent.removeExtra(DO_REFRESH);
        resendIntent.removeExtra(UPDATE_PERIOD);

        resendIntent.putExtra(UPDATE_PERIOD, NEVER_UPDATE_PERIOD);

        return resendIntent;
    }

    private boolean shouldUpdate(long updatePeriod) {
        if (updatePeriod == ALWAYS_UPDATE_PERIOD) {
            LOG.debug("shouldUpdate() : recommend update, as updatePeriod is set to ALWAYS_UPDATE");
            return true;
        }
        if (updatePeriod == NEVER_UPDATE_PERIOD) {
            LOG.debug("shouldUpdate() : recommend no update, as updatePeriod is set to NEVER_UPDATE");
            return false;
        }

        long lastUpdate = getLastUpdate();
        boolean shouldUpdate = lastUpdate + updatePeriod < System.currentTimeMillis();

        LOG.debug("shouldUpdate() : recommend {} update (lastUpdate: {}, updatePeriod: {} min)", (!shouldUpdate ? "no " : "to"), toReadable(lastUpdate), (updatePeriod / 1000 / 60));

        return shouldUpdate;
    }

    public long getLastUpdate() {
        return sharedPreferencesService.getSharedPreferences(PREFERENCES_NAME).getLong(LAST_UPDATE_PROPERTY, 0L);
    }

    public ArrayList<String> getAvailableDeviceNames() {
        ArrayList<String> deviceNames = newArrayList();
        RoomDeviceList allRoomsDeviceList = getAllRoomsDeviceList();

        for (Device device : allRoomsDeviceList.getAllDevices()) {
            deviceNames.add(device.getName() + "|" +
                    emptyOrValue(device.getAlias()) + "|" +
                    emptyOrValue(device.getWidgetName()));
        }

        return deviceNames;
    }

    private String emptyOrValue(String value) {
        if (value == null) return "";
        return value;
    }

    /**
     * Retrieves a list of all room names.
     *
     * @return list of all room names
     */
    public ArrayList<String> getRoomNameList() {
        RoomDeviceList roomDeviceList = getRoomDeviceList();
        if (roomDeviceList == null) return newArrayList();

        Set<String> roomNames = Sets.newHashSet();
        for (Device device : roomDeviceList.getAllDevices()) {
            DeviceType type = getDeviceTypeFor(device);
            if (device.isSupported() && connectionService.mayShowInCurrentConnectionType(type) && type != AT) {
                @SuppressWarnings("unchecked")
                List<String> deviceRooms = device.getRooms();
                roomNames.addAll(deviceRooms);
            }
        }
        roomNames.removeAll(roomDeviceList.getHiddenRooms());

        FHEMWEBDevice fhemwebDevice = roomListHolderService.findFHEMWEBDevice(roomDeviceList);
        return sortRooms(roomNames, fhemwebDevice);
    }

    ArrayList<String> sortRooms(Set<String> roomNames, FHEMWEBDevice fhemwebDevice) {
        final List<String> sortRooms = newArrayList();
        if (fhemwebDevice != null && fhemwebDevice.getSortRooms() != null) {
            sortRooms.addAll(Arrays.asList(fhemwebDevice.getSortRooms().split(SORT_ROOMS_DELIMITER)));
        }
        ArrayList<String> roomNamesCopy = newArrayList(roomNames);
        sort(roomNamesCopy, sortRoomsComparator(sortRooms));
        return roomNamesCopy;
    }

    private Comparator<String> sortRoomsComparator(final List<String> sortRooms) {
        return new Comparator<String>() {
            @Override
            public int compare(String lhs, String rhs) {
                int lhsIndex = sortRooms.indexOf(lhs);
                int rhsIndex = sortRooms.indexOf(rhs);

                if (lhsIndex == rhsIndex && lhsIndex == -1) {
                    // both not in sort list, compare based on names
                    return lhs.compareTo(rhs);
                } else if (lhsIndex != rhsIndex && lhsIndex != -1 && rhsIndex != -1) {
                    // both in sort list, compare indexes
                    return ((Integer) lhsIndex).compareTo(rhsIndex);
                } else if (lhsIndex == -1) {
                    // lhs not in sort list, rhs in sort list
                    return 1;
                } else {
                    // rhs not in sort list, lhs in sort list
                    return -1;
                }
            }
        };
    }

    /**
     * Retrieves the {@link RoomDeviceList} for a specific room name.
     *
     * @param roomName room name used for searching.
     * @return found {@link RoomDeviceList} or null
     */
    public RoomDeviceList getDeviceListForRoom(String roomName) {
        RoomDeviceList roomDeviceList = new RoomDeviceList(roomName);

        RoomDeviceList allRoomDeviceList = getRoomDeviceList();
        if (allRoomDeviceList != null) {
            for (Device device : allRoomDeviceList.getAllDevices()) {
                if (device.isInRoom(roomName)) {
                    roomDeviceList.addDevice(device);
                }
            }
            roomDeviceList.setHiddenGroups(allRoomDeviceList.getHiddenGroups());
            roomDeviceList.setHiddenRooms(allRoomDeviceList.getHiddenRooms());
        }

        return roomDeviceList;
    }

    public enum RemoteUpdateRequired {
        REQUIRED, NOT_REQUIRED
    }
}

File: src/main/java/li/klass/fhem/service/room/RoomListService.kt
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.service.room

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.ResultReceiver
import com.google.common.base.Optional
import com.google.common.collect.Lists.newArrayList
import com.google.common.collect.Sets
import li.klass.fhem.appwidget.service.AppWidgetUpdateService
import li.klass.fhem.constants.Actions
import li.klass.fhem.constants.Actions.*
import li.klass.fhem.constants.BundleExtraKeys.*
import li.klass.fhem.constants.PreferenceKeys
import li.klass.fhem.constants.ResultCodes
import li.klass.fhem.domain.FHEMWEBDevice
import li.klass.fhem.domain.core.DeviceType.AT
import li.klass.fhem.domain.core.DeviceType.getDeviceTypeFor
import li.klass.fhem.domain.core.FhemDevice
import li.klass.fhem.domain.core.RoomDeviceList
import li.klass.fhem.service.AbstractService
import li.klass.fhem.service.connection.ConnectionService
import li.klass.fhem.service.intent.NotificationIntentService
import li.klass.fhem.util.ApplicationProperties
import li.klass.fhem.util.DateFormatUtil.toReadable
import org.slf4j.LoggerFactory
import java.io.Serializable
import java.util.*
import java.util.Collections.sort
import java.util.concurrent.atomic.AtomicBoolean
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RoomListService @Inject
constructor() : AbstractService() {

    private val remoteUpdateInProgress = AtomicBoolean(false)

    private var resendIntents: MutableList<Intent> = newArrayList()

    @Inject
    lateinit var connectionService: ConnectionService

    @Inject
    lateinit var deviceListParser: DeviceListParser

    @Inject
    lateinit var applicationProperties: ApplicationProperties

    @Inject
    lateinit var roomListHolderService: RoomListHolderService

    @Inject
    lateinit var roomListUpdateService: RoomListUpdateService

    fun parseReceivedDeviceStateMap(deviceName: String, updateMap: Map<String, String>,
                                    vibrateUponNotification: Boolean, context: Context) {

        val deviceOptional = getDeviceForName<FhemDevice>(deviceName, Optional.absent<String>(), context)
        if (!deviceOptional.isPresent) {
            return
        }

        val device = deviceOptional.get()
        deviceListParser.fillDeviceWith(device, updateMap, context)

        LOG.info("parseReceivedDeviceStateMap()  : updated {} with {} new values!", device.name, updateMap.size)

        context.startService(Intent(NOTIFICATION_TRIGGER)
                .setClass(context, NotificationIntentService::class.java)
                .putExtra(DEVICE_NAME, deviceName)
                .putExtra(DEVICE, device)
                .putExtra(UPDATE_MAP, updateMap as Serializable)
                .putExtra(VIBRATE, vibrateUponNotification))

        context.sendBroadcast(Intent(DO_UPDATE))

        val updateWidgets = applicationProperties.getBooleanSharedPreference(PreferenceKeys.GCM_WIDGET_UPDATE, false, context)
        if (updateWidgets) {
            context.startService(Intent(Actions.REDRAW_ALL_WIDGETS)
                    .setClass(context, AppWidgetUpdateService::class.java))
        }
    }

    /**
     * Looks for a device with a given name.

     * @param deviceName name of the device
     * *
     * @return found device or null
     */
    fun <T : FhemDevice> getDeviceForName(deviceName: String?, connectionId: Optional<String>, context: Context): Optional<T> {
        if (deviceName == null) {
            return Optional.absent<T>()
        }
        return Optional.fromNullable(getAllRoomsDeviceList(connectionId, context).getDeviceFor<FhemDevice>(deviceName) as T?)
    }

    /**
     * Retrieves a [RoomDeviceList] containing all devices, not only the devices of a specific room.
     * The room device list will be a copy of the actual one. Thus, any modifications will have no effect!

     * @return [RoomDeviceList] containing all devices
     */
    fun getAllRoomsDeviceList(connectionId: Optional<String>, context: Context): RoomDeviceList {
        val originalRoomDeviceList = getRoomDeviceList(connectionId, context)
        return RoomDeviceList(originalRoomDeviceList.orNull(), context)
    }

    /**
     * Loads the currently cached [li.klass.fhem.domain.core.RoomDeviceList]. If the cached
     * device list has not yet been loaded, it will be loaded from the cache object.
     *
     *
     *
     * Watch out: Any modifications will be saved within the internal representation. Don't use
     * this method from client code!

     * @return Currently cached [li.klass.fhem.domain.core.RoomDeviceList].
     * *
     * @param context context
     */
    fun getRoomDeviceList(connectionId: Optional<String>, context: Context): Optional<RoomDeviceList> {
        return roomListHolderService.getCachedRoomDeviceListMap(connectionId, context)
    }

    fun resetUpdateProgress(context: Context) {
        LOG.debug("resetUpdateProgress()")
        remoteUpdateInProgress.set(false)
        resendIntents = newArrayList<Intent>()
        sendBroadcastWithAction(DISMISS_EXECUTING_DIALOG, context)
    }

    fun updateRoomDeviceListIfRequired(updatePeriod: Long, context: Context, connectionId: String? = null, room: String? = null, deviceName: String? = null) {
        val connectionExists = connectionService.exists(connectionId, context)
        val requiresUpdate = connectionExists && shouldUpdate(updatePeriod, Optional.fromNullable(connectionId), context, deviceName != null)
        if (requiresUpdate) {
            LOG.info("updateRoomDeviceListIfRequired() - requiring update")
            if (room != null) {
                roomListUpdateService.updateRoom(room, Optional.fromNullable(connectionId), context, updateWidgets = false)
            } else if (deviceName != null) {
                roomListUpdateService.updateSingleDevice(deviceName, Optional.fromNullable(connectionId), context, updateWidgets = false)
            } else {
                roomListUpdateService.updateAllDevices(Optional.fromNullable(connectionId), context, updateWidgets = false)
            }
        }
    }

    private fun answerError(resendIntent: Intent) {
        val receiver = resendIntent.getParcelableExtra<ResultReceiver>(RESULT_RECEIVER)
        receiver?.send(ResultCodes.ERROR, Bundle())
    }

    private fun resend(intent: Intent, context: Context) {
        LOG.info("resend() : resending {}", intent.action)

        if (intent.getIntExtra(RESEND_TRY, 0) > 2) {
            if (intent.hasExtra(RESULT_RECEIVER)) {
                val receiver = intent.getParcelableExtra<ResultReceiver>(RESULT_RECEIVER)
                receiver.send(ResultCodes.ERROR, Bundle())
                LOG.error("resend() - exceeds maximum attempts, sending error")
            }
        } else {
            context.startService(intent)
        }
    }

    private fun createResendIntent(intent: Intent): Intent {
        val resendIntent = Intent(intent)
        resendIntent.removeExtra(DO_REFRESH)
        resendIntent.removeExtra(UPDATE_PERIOD)

        resendIntent.putExtra(UPDATE_PERIOD, NEVER_UPDATE_PERIOD)
        resendIntent.putExtra(RESEND_TRY, intent.getIntExtra(RESEND_TRY, 0) + 1)

        return resendIntent
    }

    private fun shouldUpdate(updatePeriod: Long, connectionId: Optional<String>, context: Context, hasDevice: Boolean): Boolean {
        if (updatePeriod == ALWAYS_UPDATE_PERIOD) {
            LOG.debug("shouldUpdate() : recommend update, as updatePeriod is set to ALWAYS_UPDATE")
            return true
        }
        if (updatePeriod == NEVER_UPDATE_PERIOD) {
            LOG.debug("shouldUpdate() : recommend no update, as updatePeriod is set to NEVER_UPDATE")
            return false
        } else if (hasDevice) {
            LOG.debug("shouldUpdate() : has explicit device => update always")
            return true
        }

        val lastUpdate = getLastUpdate(connectionId, context)
        val shouldUpdate = lastUpdate + updatePeriod < System.currentTimeMillis()

        LOG.debug("shouldUpdate() : recommend {} update (lastUpdate: {}, updatePeriod: {} min)", if (!shouldUpdate) "no " else "to", toReadable(lastUpdate), updatePeriod / 1000 / 60)

        return shouldUpdate
    }

    fun getLastUpdate(connectionId: Optional<String>, context: Context): Long {
        return roomListHolderService.getLastUpdate(connectionId, context)
    }

    fun getAvailableDeviceNames(connectionId: Optional<String>, context: Context): ArrayList<String> {
        val deviceNames = newArrayList<String>()
        val allRoomsDeviceList = getAllRoomsDeviceList(connectionId, context)

        for (device in allRoomsDeviceList.allDevices) {
            deviceNames.add(device.name + "|" +
                    emptyOrValue(device.alias) + "|" +
                    emptyOrValue(device.widgetName))
        }

        return deviceNames
    }

    private fun emptyOrValue(value: String?): String {
        if (value == null) return ""
        return value
    }

    /**
     * Retrieves a list of all room names.

     * @param context context
     * *
     * @return list of all room names
     */
    fun getRoomNameList(connectionId: Optional<String>, context: Context): ArrayList<String> {
        val roomDeviceList = getRoomDeviceList(connectionId, context)
        if (!roomDeviceList.isPresent) return newArrayList()

        val roomNames = Sets.newHashSet<String>()
        for (device in roomDeviceList.get().allDevices) {
            val type = getDeviceTypeFor(device) ?: continue
            if (device.isSupported && connectionService.mayShowInCurrentConnectionType(type, context) && type != AT) {

                roomNames.addAll(device.rooms)
            }
        }
        roomNames.removeAll(roomDeviceList.get().hiddenRooms)

        val fhemwebDevice = roomListHolderService.findFHEMWEBDevice(roomDeviceList.get(), context)
        return sortRooms(roomNames, fhemwebDevice)
    }

    private fun sortRooms(roomNames: Set<String>, fhemwebDevice: FHEMWEBDevice?): ArrayList<String> {
        val sortRooms = newArrayList<String>()
        if (fhemwebDevice != null && fhemwebDevice.sortRooms != null) {
            sortRooms.addAll(Arrays.asList(*fhemwebDevice.sortRooms.split(SORT_ROOMS_DELIMITER.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()))
        }
        val roomNamesCopy = newArrayList(roomNames)
        sort(roomNamesCopy, sortRoomsComparator(sortRooms))
        return roomNamesCopy
    }

    private fun sortRoomsComparator(sortRooms: List<String>): Comparator<String> {
        return Comparator { lhs, rhs ->
            val lhsIndex = sortRooms.indexOf(lhs)
            val rhsIndex = sortRooms.indexOf(rhs)

            if (lhsIndex == rhsIndex && lhsIndex == -1) {
                // both not in sort list, compare based on names
                lhs.compareTo(rhs)
            } else if (lhsIndex != rhsIndex && lhsIndex != -1 && rhsIndex != -1) {
                // both in sort list, compare indexes
                lhsIndex.compareTo(rhsIndex)
            } else if (lhsIndex == -1) {
                // lhs not in sort list, rhs in sort list
                1
            } else {
                // rhs not in sort list, lhs in sort list
                -1
            }
        }
    }

    /**
     * Retrieves the [RoomDeviceList] for a specific room name.

     * @param roomName room name used for searching.
     * *
     * @return found [RoomDeviceList] or null
     */
    fun getDeviceListForRoom(roomName: String, connectionId: Optional<String>, context: Context): RoomDeviceList {
        val roomDeviceList = RoomDeviceList(roomName)

        val allRoomDeviceList = getRoomDeviceList(connectionId, context)
        if (allRoomDeviceList.isPresent) {
            for (device in allRoomDeviceList.get().allDevices) {
                if (device.isInRoom(roomName)) {
                    roomDeviceList.addDevice(device, context)
                }
            }
            roomDeviceList.hiddenGroups = allRoomDeviceList.get().hiddenGroups
            roomDeviceList.hiddenRooms = allRoomDeviceList.get().hiddenRooms
        }

        return roomDeviceList
    }

    enum class RemoteUpdateRequired {
        REQUIRED, NOT_REQUIRED
    }

    companion object {

        private val LOG = LoggerFactory.getLogger(RoomListService::class.java)

        val PREFERENCES_NAME = RoomListService::class.java.name

        val LAST_UPDATE_PROPERTY = "LAST_UPDATE"

        val NEVER_UPDATE_PERIOD: Long = 0
        val ALWAYS_UPDATE_PERIOD: Long = -1

        val SORT_ROOMS_DELIMITER = " "
    }
}
